Provide web access to media files received from devices

Abyss777 opened the pull request:
There is a switch `media.dirAllowed` to allow access directories. Admin can enumerate root directory, users only device directories they have access.

![image](https://user-images.githubusercontent.com/5688080/35671023-f405215c-075b-11e8-870d-415f10adbe3b.png)

- Added `DefaultServlet` to serve folder with media files as static content
- Added filter to check access to device

fix #3688 

More comments in code...

Abyss777 commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -0,0 +1,104 @@
+            HttpSession session = ((HttpServletRequest) request).getSession(false);
+            Long userId = null;
+            if (session != null) {
+                userId = (Long) session.getAttribute(SessionResource.USER_ID_KEY);
```
Authorization by session only is supported

Abyss777 commented on src/org/traccar/web/WebServer.java:
```diff
@@ -156,6 +161,19 @@ private void initOldWebApp() {
 
+    private void initMedia() {
+        ServletContextHandler servletHandler = new ServletContextHandler(ServletContextHandler.SESSIONS);
+        servletHandler.setContextPath("/api/media");
```
We have to use sub path to make SessionManager use same sessions as for `/api`

tananaev replied:
Can we add it to the API ServletContextHandler?

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        dirAllowed = Context.getConfig().getBoolean("media.dirAllowed");
```
I would prefer to use full words. For example `directoryAllow`.

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -0,0 +1,104 @@
+        dirAllowed = Context.getConfig().getBoolean("media.dirAllowed");
+    }
+
+    private static void formatError(HttpServletResponse response, Exception e) throws IOException {
```
Can we set status and return directly in the `doFilter` method?

Abyss777 replied:
Sure

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -0,0 +1,104 @@
+                } else {
+                    throw new SecurityException("Wrong path");
+                }
+            } else if (parts.length == 2 && !dirAllowed) {
```
What is this check for?

Abyss777 replied:
It is to prevent list directory for ordinary users, but just thought may be it is not needed, because it is forbidden on servlet level anyway.

Abyss777 commented:
- Move media servlet to api context (it is much better, thanks for the tip)
- Removed any directory check from filter, let servlet handle it
- Renamed switch to `media.directoryAllow`
- Moved error formatting back to `doFilter`
- Some other cleanup

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -93,7 +72,17 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         } catch (Exception e) {
-            formatError((HttpServletResponse) response, e);
+            HttpServletResponse httpResponse = (HttpServletResponse) response;
+            if (e instanceof SecurityException) {
```
Actually what I meant is that instead of throwing exceptions and catching them you can just set status directly in most cases.

Abyss777 replied:
We still need catch Security exceptions when checking user and device. Just made it in common way.

tananaev replied:
Yes, but do we need to catch anything else other than security? It seems like a lot of overhead. And even if we do, `instanceof` is not the best option. You can just have separate catch clauses. But I think if we don't need exception we shouldn't use it.

Abyss777 commented:
You are right. Removed throwing of exceptions. Used `DeviceManager` instead identity to `getByUniqueId` to catch only `SQLException`

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -54,34 +54,32 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
             if (userId == null) {
-                throw new NotAuthorizedException("Not authorized");
+                ((HttpServletResponse) response).setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+                response.getWriter().println("Not authorized");
```
Do we really need to write some text in the output? I think it makes sense for exception only. Hardcoded text doesn't provide any value over status code.

tananaev commented on src/org/traccar/api/MediaFilter.java:
```diff
@@ -54,34 +54,32 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
             }
             if (userId == null) {
-                throw new NotAuthorizedException("Not authorized");
+                ((HttpServletResponse) response).setStatus(HttpServletResponse.SC_UNAUTHORIZED);
```
You can cast it once at the beginning of the method. It will probably make code easier to read.

tananaev commented:
Merged, thanks.

Abyss777 commented:
Thanks. Sorry for so much corrections in this PR

tananaev commented:
One more thing I though of. What would happen if media folder is not specified?

Abyss777 commented:
Good question. It returns 404 for me, but I think it tries open some folder. But do not know which one, neither `./api/media` /  `./media` or `/api/media` /  `/media`

Abyss777 commented:
What about choose some default folder? like `./media` ?

tananaev commented:
I think it would be better to allow disabling of this functionality.
